azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr, PriorityQueue
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "TelemetryServiceHandler"
+    _MAX_TIMEOUT = datetime.timedelta(minutes=5).seconds
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
+
+    @staticmethod
+    def get_thread_name():
+        return TelemetryServiceHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event_func(self, event):
+        # Add event to queue and set event
+        self._queue.put(event)
+
+        # Set the event if any enqueue happens (even if already set) to trigger sending those events
+        self._should_process_events.set()
+
+    def _process_telemetry_thread(self):
+        logger.info("Successfully started the {0} thread".format(self.get_thread_name()))
+        try:
+            # On demand wait, start processing as soon as there is any data available in the queue
+            # In worst case, also keep checking every 5 mins to ensure that no data is being missed
+            while not self.stopped():
+                self._should_process_events.wait(timeout=TelemetryServiceHandler._MAX_TIMEOUT)
```

Comment:
Previously we had a bug where the agent would hang on service stop when it was doing a call to wait() on the extension's processes. Nor sure if Event would have the same problem? Could you trying issuing a service stop while this call is blocked waiting for the timeout?

Reply:
Definitively watch out for that call to wait(). The documentation for Event doesn't mention anything, but the documentation for Queue.get() points out that this function is not interruptible (and, unfortunately, that is distro-dependent):

`Prior to 3.0 on POSIX systems, and for all versions on Windows, if block is true and timeout is None, this operation goes into an uninterruptible wait on an underlying lock. This means that no exceptions can occur, and in particular a SIGINT will not trigger a KeyboardInterrupt.
`

Reply:
Good point. Though I think if we get a SIGTERM we just `sys.exit` so I dont think any thread would matter at that point. 

```python
        if self.signal_handler is None:
            self.signal_handler = signal.signal(signal.SIGTERM, self.forward_signal)

    def forward_signal(self, signum, frame):
        if signum == signal.SIGTERM:
            self._shutdown()

        if self.child_process is None:
            return

 def _shutdown(self):
        self.running = False

        if not os.path.isfile(self._sentinel_file_path()):
            return
```

Reply:
If the wait is uninterruptible it won't respond to signals (other than KILL). We used to have a hang when trying to stop the agent service because there was a thread on wait() for extension's output.

Reply:
Ahh interesting, didnt know that. Then do you think its better to do the `while()..sleep` approach here too? I was trying to avoid that because it wont be an on demand thing then but these other issues are a pain too

Reply:
Our bug was a hang while updating the agent when an extension's forked process was still running and we were waiting for its output (guess which extension it was :))

I learnt about uninterruptible waits while checking out how the cpu scheduler works (for the cgroups stuff) :)

The cgroups stuff made read how the cpu scheduler works, that is where I learnt about those waits :)

Reply:
That's fair, even small interrupts should be stopped, not just SIGTERMS. I'm guessing the thread.sleep() are interruptible waits and can be broken if the process wants to kill.

Though would something like this work, if we setup a signal trapper for each signal and stop() each thread specifically rather than just killing them - 

Example: 
```python
for i in [x for x in dir(signal) if x.startswith("SIG")]:
  try:
    signum = getattr(signal,i)
    signal.signal(signum, stop_all_threads())
  except (OSError, RuntimeError) as m: #OSError for Python3, RuntimeError for 2
    print ("Skipping {}".format(i))
```

Reply:
As per our discussion, reducing the max timeout for wait to 5 secs to ensure we dont hang for long